File: src/voicetype/asr.py
"""Speech-to-text via OpenAI-compatible API."""

import logging
import time

from voicetype.api_client import ApiClient
from voicetype.config import AppConfig
from voicetype.retry import retry_call

logger = logging.getLogger(__name__)

# Prompt hint for auto-detect mode to improve Chinese-English mixed recognition.
# Whisper uses the prompt parameter as context to bias language detection.
_AUTO_DETECT_PROMPT = "以下是普通话和English混合的句子。"


class Transcriber:
    def __init__(self, config: AppConfig):
        self.config = config
        self._client = ApiClient(
            api_key=config.asr.api_key,
            base_url=config.asr.base_url,
            timeout=30,
        ).client

    def transcribe(self, audio_path: str) -> str:
        """Transcribe audio file to text."""
        kwargs = {
            "model": self.config.asr.model,
        }
        lang = self.config.asr.language
        if lang and lang != "auto":
            kwargs["language"] = lang
        else:
            kwargs["prompt"] = _AUTO_DETECT_PROMPT

        logger.debug("Starting transcription: model=%s, lang=%s", kwargs["model"], lang)

        def _call():
            # Re-open the file per attempt: a failed request may have read
            # part of the stream, leaving the file pointer mid-file.
            with open(audio_path, "rb") as f:
                return self._client.audio.transcriptions.create(file=f, **kwargs)

        # Retry on transient failures (connection drops, 429, 5xx) so a
        # single hiccup doesn't discard the user's recorded audio.
        start = time.monotonic()
        try:
            response = retry_call(_call)
        except Exception as e:
            logger.error(
                "Transcription failed after retries: %s (%.1fs)",
                e,
                time.monotonic() - start,
                exc_info=True,
            )
            raise
        text = response.text.strip()
        elapsed = time.monotonic() - start
        logger.info("Transcription done in %.1fs: %d chars", elapsed, len(text))
        if not text:
            logger.warning("Transcription returned empty text")
        return text

File: desktop/src/main/services/asr.ts
// Batch speech-to-text via an OpenAI-compatible /audio/transcriptions API —
// port of voicetype/asr.py + api_client.py. Uses fetch with multipart FormData
// instead of the openai SDK; retry/backoff semantics are identical.

import { HttpError, retryCall } from './retry'
import type { AppConfig } from '../../shared/types'

// Prompt hint for auto-detect mode to improve Chinese-English mixed recognition.
const AUTO_DETECT_PROMPT = '以下是普通话和English混合的句子。'

async function fetchWithTimeout(url: string, init: RequestInit, timeoutMs: number): Promise<Response> {
  const controller = new AbortController()
  const timer = setTimeout(() => controller.abort(), timeoutMs)
  try {
    return await fetch(url, { ...init, signal: controller.signal })
  } finally {
    clearTimeout(timer)
  }
}

/** Fetch the provider's model list (GET /models) for the settings refresh button. */
export async function fetchModels(apiKey: string, baseUrl: string, timeoutMs = 10_000): Promise<string[]> {
  const base = baseUrl.replace(/\/+$/, '')
  const res = await fetchWithTimeout(
    `${base}/models`,
    { headers: { Authorization: `Bearer ${apiKey}` } },
    timeoutMs
  )
  if (!res.ok) throw new HttpError(res.status, `HTTP ${res.status}`)
  const data = (await res.json()) as { data?: Array<{ id?: string }> }
  return (data.data ?? [])
    .map((m) => m.id ?? '')
    .filter(Boolean)
    .sort((a, b) => a.localeCompare(b))
}

export class Transcriber {
  private readonly apiKey: string
  private readonly baseUrl: string
  private readonly model: string
  private readonly language: string

  constructor(config: AppConfig) {
    this.apiKey = config.asr.api_key
    this.baseUrl = config.asr.base_url.replace(/\/+$/, '')
    this.model = config.asr.model
    this.language = config.asr.language
  }

  /** Best-effort connection warmup so the first real call skips TLS setup. */
  async warmup(): Promise<void> {
    try {
      await fetchModels(this.apiKey, this.baseUrl, 4000)
    } catch {
      // swallow — warmup is opportunistic
    }
  }

  async transcribe(wav: Buffer, sampleRate: number): Promise<string> {
    const form = () => {
      const fd = new FormData()
      // Rebuild the Blob per attempt: a failed request may have consumed part
      // of the stream (mirrors asr.py reopening the file per attempt).
      fd.append('file', new Blob([new Uint8Array(wav)], { type: 'audio/wav' }), 'recording.wav')
      fd.append('model', this.model)
      const lang = this.language
      if (lang && lang !== 'auto') {
        fd.append('language', lang)
      } else {
        fd.append('prompt', AUTO_DETECT_PROMPT)
      }
      // Some providers infer chunk duration from these optional fields.
      fd.append('response_format', 'json')
      void sampleRate
      return fd
    }

    const text = await retryCall(async () => {
      const res = await fetchWithTimeout(
        `${this.baseUrl}/audio/transcriptions`,
        { method: 'POST', headers: { Authorization: `Bearer ${this.apiKey}` }, body: form() },
        30_000
      )
      if (!res.ok) throw new HttpError(res.status, `transcription HTTP ${res.status}`)
      const data = (await res.json()) as { text?: string }
      return data.text ?? ''
    })

    const trimmed = text.trim()
    return trimmed
  }
}
